firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -103,4 +124,105 @@ String getApplicationId() {
   }
+
+  /**
+   * Create firebase installation id for the {@link FirebaseApp} on FIS Servers and client side
+   * cache.
+   *
+   * <pre>
+   *     The workflow is:
+   *         check if cache empty or cache status is REGISTER_ERROR
+   *                                 |
+   *                           create random fid
+   *                                 |
+   *               update cache with cache status UNREGISTERED
+   *                                 |
+   *                           send http request to backend
+   *                                 |                  |
+   *  on success: set cache entry status to REGISTERED  |
+   *                                 |                  |
+   *                                 |         on failure: set cache entry status to REGISTER_ERROR
+   *                                 |                     |
+   *                                return          throw exception
+   *
+   *
+   *                      else if cached FID exists
+   *                             |            |
+   *     if cache status is UNREGISTERED      |
+   *                  |                   return cached FID
+   *     send http request to backend
+   *               |               |
+   *  on success: set cache        \
+   *  entry status to REGISTERED   \
+   *                |              \
+   *                |         on failure: set cache entry status to REGISTER_ERROR
+   *                |                     |
+   *              return          throw exception
+   * </pre>
+   */
+  @WorkerThread
+  private String createFirebaseInstallationId() throws FirebaseInstallationsException {
+
+    FiidCacheEntryValue cacheEntryValue = localCache.readCacheEntryValue();
+
+    if (cacheEntryValue == null
+        || cacheEntryValue.getCacheStatus() == FiidCache.CacheStatus.REGISTER_ERROR) {
+      String fid = Utils.createRandomFid();
+
+      boolean firstUpdateCacheResult =
+          localCache.insertOrUpdateCacheEntry(
+              FiidCacheEntryValue.create(fid, FiidCache.CacheStatus.UNREGISTERED, "", "", 0, 0));
+
+      if (!firstUpdateCacheResult) {
+        throw new FirebaseInstallationsException(
+            "Failed to update client side cache.",
+            FirebaseInstallationsException.Status.CLIENT_ERROR);
+      }
+
+      registerAndSaveFID(fid);
+
+      return fid;
+    } else if (!cacheEntryValue.getFirebaseInstallationId().isEmpty()) {
+
+      if (cacheEntryValue.getCacheStatus() == FiidCache.CacheStatus.UNREGISTERED) {
+        registerAndSaveFID(cacheEntryValue.getFirebaseInstallationId());
+      }
+
+      return cacheEntryValue.getFirebaseInstallationId();
+    }
+    return null;
+  }
+
+  /**
+   * Registers the created FID with FIS Servers if the Network is available and update the cache.
+   */
+  private void registerAndSaveFID(String fid) throws FirebaseInstallationsException {
+    try {
+      if (Utils.isNetworkAvailable(firebaseApp.getApplicationContext())) {
```

Comment:
I would strongly recommend to avoid large if-clauses like this one.

How about the first line of the method is this:
if (! Utils.isNetworkAvailable(firebaseApp.getApplicationContext())) {
  return;
}

Reply:
Also: Why is this if-clause part of the try-catch block, can we pull it before the try-catch block?
Or does #isNetworkAvailable throw an exception?

Reply:
Updated. Thanks for pointing it out.